src/transformers/models/mistral3/modular_mistral3.py

```diff
@@ -0,0 +1,451 @@
+        """
+        # check if images and text inputs are reversed for BC
+        images, text = _validate_images_text_input_order(images, text)
+
+        output_kwargs = self._merge_kwargs(
+            Mistral3ProcessorKwargs,
+            tokenizer_init_kwargs=self.tokenizer.init_kwargs,
+            **kwargs,
+        )
+
+        if images is not None:
+            if is_image_or_image_url(images):
+                images = [images]
+            elif isinstance(images, (list, tuple)) and is_image_or_image_url(images[0]):
+                pass
+            elif (
+                isinstance(images, (list, tuple))
+                and isinstance(images[0], (list, tuple))
+                and is_image_or_image_url(images[0][0])
+            ):
+                images = [image for sublist in images for image in sublist]
+            else:
+                raise ValueError(
+                    "Invalid input images. Please provide a single image, a list of images, or a list of lists of images."
+                )
+            images = [load_image(im) if isinstance(im, str) else im for im in images]
+            image_inputs = self.image_processor(images, patch_size=self.patch_size, **output_kwargs["images_kwargs"])
+        else:
+            image_inputs = {}
+
+        if isinstance(text, str):
+            text = [text]
+        elif not isinstance(text, list) and not isinstance(text[0], str):
+            raise ValueError("Invalid input text. Please provide a string, or a list of strings")
+
+        # try to expand inputs in processing if we have the necessary parts
+        prompt_strings = text
+        if image_inputs.get("pixel_values") is not None:
+            # Replace the image token with the expanded image token sequence
+            image_sizes = iter(image_inputs["image_sizes"])
+            prompt_strings = []
+            replace_strings = []
+
+            for sample in text:
+                while self.image_token in sample:
+                    height, width = next(image_sizes)
+                    num_height_tokens = height // (self.patch_size * self.spatial_merge_size)  # main diff with Pixtral
+                    num_width_tokens = width // (self.patch_size * self.spatial_merge_size)  # main diff with Pixtral
+                    replace_tokens = [
+                        [self.image_token] * num_width_tokens + [self.image_break_token]
+                    ] * num_height_tokens
+                    # Flatten list
+                    replace_tokens = [item for sublist in replace_tokens for item in sublist]
+                    replace_tokens[-1] = self.image_end_token
+                    replace_str = "".join(replace_tokens)
+                    replace_strings.append(replace_str)
+                    sample = sample.replace(self.image_token, "<placeholder>", 1)
+
+                while "<placeholder>" in sample:
```

Comment:
always worries me a bit to add placeholder tokens a strings before the tokenization and not after, but think that's out of scope here haha